addon/-private/properties/collection/main.js

```diff
@@ -0,0 +1,119 @@
+/* global Symbol */
+import Ember from 'ember';
+import { buildSelector, assign } from '../../helpers';
+import { create } from '../../create';
+import {
+  ELEMENT_NOT_FOUND,
+  throwBetterError
+} from '../../better-errors';
+import { count } from '../count';
+import Ceibo from 'ceibo';
+
+export class Collection {
+  constructor(definition, parent, key) {
+    this.definition = definition;
+    this.parent = parent;
+    this.key = key;
+
+    this._itemCounter = create({
+      count: count(definition.scope, {
+        resetScope: definition.resetScope,
+        testContainer: definition.testContainer
+      })
+    }, { parent });
+
+    this._items = [];
+  }
+
+  get length() {
+    return this._itemCounter.count;
+  }
+
+  objectAt(index) {
+    if (typeof this._items[index] === 'undefined') {
+      let { definition, parent, key } = this;
+      let scope = buildSelector({}, definition.scope, { at: index });
+
+      let finalizedDefinition = assign({}, definition);
+
+      finalizedDefinition.scope = scope;
+
+      let tree = create(finalizedDefinition, { parent });
+
+      // Change the key of the root node
+      Ceibo.meta(tree).key = `${key}[${index}]`;
+
+      this._items[index] = tree;
+    }
+
+    let itemPage = this._items[index];
+
+    if (index > this.length - 1) {
+      throwBetterError(
+        itemPage,
+        null,
```

Comment:
Currently, this gives us an error like this:
```
Error: Element not found.  
  PageObject: 'page.foo.bar[1].'   
  Selector: '[data-test-bar]:eq(1)'
```

But this might be a bit nicer:

```
Error: Element not found.
  PageObject: 'page.foo.objectAt(1)'
  Selector: '[data-test-bar]:eq(1)'
```

So suggest this:
```javascript
throwBetterError(
  this.parent,
  `objectAt(${index})`,
  ELEMENT_NOT_FOUND,
  { selector: itemPage.scope }
);
```